aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp

```diff
@@ -125,19 +131,117 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
   }
-#else // USE_FBGEMM
-  Tensor operator()(
-      Tensor, /* weight */
-      c10::optional<Tensor>, /* bias */
-      torch::List<int64_t>, /* stride */
-      torch::List<int64_t>, /* padding */
-      torch::List<int64_t>, /* dilation */
-      int64_t /* groups */
-  ) {
+#endif // USE_FBGEMM
+#ifdef USE_PYTORCH_QNNPACK
+  at::Tensor qnnpack_conv_prepack(
+      Tensor weight,
+      c10::optional<Tensor> bias_in,
+      torch::List<int64_t> stride,
+      torch::List<int64_t> padding,
+      torch::List<int64_t> dilation,
+      int64_t groups) {
+    TORCH_CHECK(
+        weight.ndimension() == 4,
+        "quantized::conv_prepack (qnnpack): Weights are expected to have 4 dimensions");
+    const auto qtype = weight.qscheme();
+    TORCH_CHECK(
+        weight.qscheme() == kPerTensorAffine,
+        "quantized::conv_prepack (qnnpack): only supports Per Tensor Quantization Scheme")
+    TORCH_CHECK(
+        stride.size() == 2,
+        "quantized::conv_prepack (qnnpack): 2D convolution only");
+    TORCH_CHECK(
+        padding.size() == 2,
+        "quantized::conv_prepack (qnnpack): Specify top/left padding only. \
+       bottom/right padding assumed to be equal to top/left");
+    TORCH_CHECK(
+        dilation.size() == 2,
+        " quantized::conv_prepack (qnnpack): 2D convolution only");
+
+    initQNNPACK();
+
+    // QNNPACK expects weights to be of the format {out_c, kH, kW, in_c/groups}
+    const size_t out_ch = weight.size(0);
+    const uint32_t kernel_h = weight.size(1);
+    const uint32_t kernel_w = weight.size(2);
+    const size_t in_ch = weight.size(3) * groups;
+
+    Tensor bias;
+    if (bias_in.has_value()) {
+      bias = bias_in.value();
+    } else {
+      bias = at::empty(out_ch, at::kFloat);
+      bias = at::quantize_linear(bias, 1.0, 0, kQInt32);
+    }
     TORCH_CHECK(
-        false, "This PyTorch installation was not built with FBGEMM operators");
+        !bias.defined() || (bias.ndimension() == 1 && bias.size(0) == out_ch),
+        "quantized::conv_prepack (qnnpack): expected bias to be 1-dimensional with ",
+        out_ch,
+        " elements",
+        ", but got bias of size ",
+        bias.sizes(),
+        " instead");
+
+    uint32_t stride_h = stride[0];
+    uint32_t stride_w = stride[1];
+    uint32_t pad_t = padding[0];
+    uint32_t pad_l = padding[1];
+    uint32_t dilation_h = dilation[0];
+    uint32_t dilation_w = dilation[1];
+
+    qnnpack::conv_param_t conv_p(
+        {kernel_w, kernel_h},
+        {stride_w, stride_h},
+        {dilation_w, dilation_h},
+        {pad_t, pad_l, pad_t, pad_l},
+        groups,
+        in_ch,
+        out_ch,
+        weight.q_zero_point(),
+        weight.q_scale(),
+        std::numeric_limits<uint8_t>::min(),
+        std::numeric_limits<uint8_t>::max());
+
+    auto weight_contig = weight.contiguous();
+    auto bias_contig = bias.contiguous();
+    auto wt_ptr =
+        guts::make_unique<PackedConvWeightsQnnp>(PackedConvWeightsQnnp{
```

Comment:
Where is the weight converted from int8 to uint8?

Reply:
That is in a separate PR -https://github.com/pytorch/pytorch/pull/26307